internal/provider/services/storage/storage_test.go

```diff
@@ -0,0 +1,463 @@
+package storage
+
+import (
+	"bytes"
+	"errors"
+	mocks2 "github.com/terrariumcloud/terrarium/internal/storage/mocks"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	terrarium "github.com/terrariumcloud/terrarium/internal/provider/services"
+	"github.com/terrariumcloud/terrarium/internal/provider/services/mocks"
+	"github.com/terrariumcloud/terrarium/pkg/terrarium/provider"
+	"google.golang.org/grpc"
+)
+
+type ClosingBuffer struct {
+	*bytes.Buffer
+}
+
+func (cb *ClosingBuffer) Close() error {
+	return nil
+}
+
+// Test_RegisterStorageWithServer checks:
+// - if there was no error with bucket init
+// - if error was returned when bucket init fails
+func Test_RegisterStorageWithServer(t *testing.T) {
+	t.Parallel()
+
+	t.Run("when bucket init is successful", func(t *testing.T) {
+		s3Client := &mocks2.S3{}
+
+		ss := &StorageService{Client: s3Client}
+
+		s := grpc.NewServer(*new([]grpc.ServerOption)...)
+
+		err := ss.RegisterWithServer(s)
+
+		if err != nil {
+			t.Errorf("Expected no error, got %v.", err)
+		}
+
+		if s3Client.HeadBucketInvocations != 1 {
+			t.Errorf("Expected 1 call to HeadBucket, got %v.", s3Client.HeadBucketInvocations)
+		}
+
+		if s3Client.CreateBucketInvocations != 0 {
+			t.Errorf("Expected no calls to CreateBucket, got %v.", s3Client.CreateBucketInvocations)
+		}
+	})
+
+	t.Run("when bucket init fails", func(t *testing.T) {
+		s3Client := &mocks2.S3{
+			HeadBucketError:   errors.New("some error"),
+			CreateBucketError: errors.New("some error"),
+		}
+
+		vms := &StorageService{Client: s3Client}
+
+		s := grpc.NewServer(*new([]grpc.ServerOption)...)
+
+		err := vms.RegisterWithServer(s)
+
+		if err != BucketInitializationError {
+			t.Errorf("Expected %v, got %v.", BucketInitializationError, err)
+		}
+
+		if s3Client.HeadBucketInvocations != 1 {
+			t.Errorf("Expected 1 call to DescribeTable, got %v.", s3Client.HeadBucketInvocations)
+		}
+
+		if s3Client.CreateBucketInvocations != 1 {
+			t.Errorf("Expected 0 calls to CreateTable, got %v.", s3Client.CreateBucketInvocations)
+		}
+	})
+}
+
+// Test_DownloadProviderSourceZip checks:
+// - if correct response is returned when source zip is downloaded
+// - if error is returned when GetObject fails
+// - if error is returned when Send fails
+// - if error is returned when wrong content lenght is read
+func Test_DownloadProviderSourceZip(t *testing.T) {
+	t.Parallel()
+
+	t.Run("when source zip is downloaded", func(t *testing.T) {
+		var length int64 = 1000
```

Comment:
You're only testing a single chunk here, better to test with a larger value to ensure your chunking logic works.